src/context.rs

```diff
@@ -47,54 +46,23 @@ impl Context {
     ///
     /// `ENOSYS`    No kernel support for RDMA.
     pub(crate) fn open(dev_name: Option<&str>, port_num: u8, gid_index: usize) -> io::Result<Self> {
-        let mut num_devs: i32 = 0;
-        // ibv_get_device_list() returns the array of available RDMA devices on success, returns NULL and sets errno
-        // if the request fails. If no devices are found, then num_devices is set to 0, and non-NULL is returned.
-        // SAFETY: ffi
-        // TODO: check safety
-        let dev_list_ptr = unsafe { ibv_get_device_list(&mut num_devs) };
-        if dev_list_ptr.is_null() {
-            let err = log_ret_last_os_err_with_note("This is a basic verb that shouldn't fail, check if the module ib_uverbs is loaded.");
-            return Err(err);
+        let dev_list = log_ret(
+            DeviceList::available(),
+            "This is a basic verb that shouldn't fail, check if the module ib_uverbs is loaded.",
+        )?;
+
+        let dev = match dev_name {
+            Some(name) => dev_list.iter().find(|&d| d.name() == name),
+            None => dev_list.get(0),
         }
-        // SAFETY: ?
-        // TODO: check safety
-        let dev_list = unsafe { std::slice::from_raw_parts(dev_list_ptr, num_devs.cast()) };
-        let dev = if let Some(dev_name_inner) = dev_name {
-            dev_list
-                .iter()
-                .find(|iter_dev| -> bool {
-                    // SAFETY: ffi
-                    // TODO: check safety
-                    let name = unsafe { ibv_get_device_name(**iter_dev) };
-                    if name.is_null() {
-                        warn!("get null dev name");
-                        return false;
-                    }
-                    // SAFETY: ?
-                    // TODO: check safety
-                    let name = unsafe { CStr::from_ptr(name) }.to_str();
-                    if name.is_err() {
-                        warn!("Device name {:?} is not valid", name);
-                        return false;
-                    }
-
-                    // We've checked the name is not error
-                    #[allow(clippy::unwrap_used)]
-                    let name = name.unwrap();
-                    dev_name_inner.eq(name)
-                })
-                .ok_or(io::ErrorKind::NotFound)?
-        } else {
-            dev_list.get(0).ok_or(io::ErrorKind::NotFound)?
-        };
+        .ok_or(io::ErrorKind::NotFound)?;
+
         // SAFETY: ffi
         // TODO: check safety
```

Comment:
Do we need more checks here?
For ibv APIs, if we correctly pass a value returned by another ibv API, its safety depends on the ibv API itself.
So I think only `SAFETY: ffi` is ok here, what's your opinion?

Reply:
Added some explanations in the latest commit